authentication/grpc.go

```diff
@@ -0,0 +1,91 @@
+package authentication
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/go-kit/log"
+	grpc_middleware "github.com/grpc-ecosystem/go-grpc-middleware/v2"
+	grpc_middleware_auth "github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/auth"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+// GRPCMiddlewareFunc is a function type able to return authentication middleware for
+// a given tenant. If no middleware is found, the second return value should be false.
+type GRPCMiddlewareFunc func(tenant string) (grpc.StreamServerInterceptor, bool)
+
+// WithGRPCTenantHeader returns a new StreamServerInterceptor that adds the tenant and tenantid
+// to the stream Context.
+func WithGRPCTenantHeader(header string, tenantIDs map[string]string, logger log.Logger) grpc.StreamServerInterceptor {
+	header = strings.ToLower(header)
+
+	return func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		ctx := ss.Context()
+		md, ok := metadata.FromIncomingContext(ctx)
+		if !ok {
+			return status.Errorf(codes.Internal, "metadata not found")
+		}
+
+		headerTenants := md[header]
+		if len(headerTenants) == 0 {
+			return status.Errorf(codes.InvalidArgument, fmt.Sprintf("header %q not set", header))
+		}
+		if len(headerTenants) > 1 {
+			return status.Errorf(codes.InvalidArgument, fmt.Sprintf("header %q requested multiple tenants", header))
+		}
+
+		tenant := headerTenants[0]
+		ctx = context.WithValue(ctx, tenantKey, tenant)
+
+		id, ok := tenantIDs[tenant]
+		if !ok {
+			// Rather than letting the caller know their tenant was invalid, just record that it was invalid
+			id = "__no_tenant_id"
+		}
```

Comment:
What is the rationale behind this instead of returning error?

Reply:
Normally for security if someone doesn't have a valid `Authorization` we don't give them any hints about their message.  Such as "invalid tenant ID", which would let an unauthorized user check if a tenant existed.

Later I realized that the Observatorium token validation endpoint is different for each tenant, thus it is impossible to authorize without a valid tenant header.  Thus not failing here doesn't provide the additional security I wanted -- a user with an invalid bearer JWT can check if tenants exist.

Reply:
Got it, thanks for the explanation :+1: